safekeeper/src/timeline.rs

```diff
@@ -508,6 +510,27 @@ impl Timeline {
         }
     }
 
+    /// Pause timeline, forbidding writes to it, or resume after it has been paused.
+    pub async fn pause(&self, paused: bool) -> Result<()> {
+        {
+            let mut ss = self.write_shared_state(true).await?;
+            if paused && !ss.sk.state.paused {
+                let mut persistent_state = ss.sk.state.start_change();
+                persistent_state.paused = true;
+                ss.sk.state.finish_change(&persistent_state).await?;
```

Comment:
Not sure how prevalent this pattern is in existing safekeeper code, but in general it's a bit undesirable to do disk I/O while holding a sync mutex.  What if this function persisted the change to disk first, and then took the lock?

Reply:
Per timeline locking in safekeepers is very simple: basically any operation takes this lock to do its job. That 1) protects from concurrent modifications, like processing messages from multiple computes 2) is ok in terms of performance because during normal work the only lock acquisition which happens regularly is from walreceiver (single compute connection), it hasn't anyone to compete with. Pausing is not an exception here.